pkg/replicater/scheme.go

```diff
@@ -0,0 +1,381 @@
+		level.Debug(bf.logger).Log("msg", "filtering block", "reason", "compaction levels don't match", "got_compaction_level", gotCompactionLevel, "expected_compaction_level", expectedCompactionLevel)
+		return false
+	}
+
+	return true
+}
+
+type blockFilterFunc func(b *metadata.Meta) bool
+
+type replicationScheme struct {
+	fromBkt objstore.BucketReader
+	toBkt   objstore.Bucket
+
+	blockFilter blockFilterFunc
+
+	logger  log.Logger
+	metrics *replicationMetrics
+}
+
+type replicationMetrics struct {
+	originIterations  prometheus.Counter
+	originMetaLoads   prometheus.Counter
+	originPartialMeta prometheus.Counter
+
+	blocksAlreadyReplicated prometheus.Counter
+	blocksReplicated        prometheus.Counter
+	objectsReplicated       prometheus.Counter
+}
+
+func newReplicationMetrics(reg prometheus.Registerer) *replicationMetrics {
+	m := &replicationMetrics{
+		originIterations: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_replicate_origin_iterations_total",
+			Help: "Total number of objects iterated over in the origin bucket.",
+		}),
+		originMetaLoads: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_replicate_origin_meta_loads_total",
+			Help: "Total number of meta.json reads in the origin bucket.",
+		}),
+		originPartialMeta: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_replicate_origin_partial_meta_reads_total",
+			Help: "Total number of partial meta reads encountered.",
+		}),
+		blocksAlreadyReplicated: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_replicate_blocks_already_replicated_total",
+			Help: "Total number of blocks skipped due to already being replicated.",
+		}),
+		blocksReplicated: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_replicate_blocks_replicated_total",
+			Help: "Total number of blocks replicated.",
+		}),
+		objectsReplicated: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_replicate_objects_replicated_total",
+			Help: "Total number of objects replicated.",
+		}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(m.originIterations)
+		reg.MustRegister(m.originMetaLoads)
+		reg.MustRegister(m.originPartialMeta)
+		reg.MustRegister(m.blocksAlreadyReplicated)
+		reg.MustRegister(m.blocksReplicated)
+		reg.MustRegister(m.objectsReplicated)
+	}
+
+	return m
+}
+
+func newReplicationScheme(logger log.Logger, metrics *replicationMetrics, blockFilter blockFilterFunc, from objstore.BucketReader, to objstore.Bucket) *replicationScheme {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	return &replicationScheme{
+		logger:      logger,
+		blockFilter: blockFilter,
+		fromBkt:     from,
+		toBkt:       to,
+		metrics:     metrics,
+	}
+}
+
+func (rs *replicationScheme) execute(ctx context.Context) error {
+	availableBlocks := []*metadata.Meta{}
+
+	level.Debug(rs.logger).Log("msg", "scanning blocks available blocks for replication")
+
+	if err := rs.fromBkt.Iter(ctx, "", func(name string) error {
```

Comment:
Let's add `TODO: Use block.MetaFetcher with filters instead.` (:  Let's not do to all in the same PR!

Reply:
Also, there are some TODO in original repo, need to move all here?

Reply:
The mixin part would finish in other PR.